### ISSUE TITLE
Support distribution of GCC on OSX

### ISSUE THREAD

**Issue by jamienoss** [ContinuumIO/anaconda-issues#10081]
Publish distribution of GCC for OSX

I've opened this issue as a separation of the following issue https://github.com/conda/conda/issues/7775

Some of the relevant comments there are:

https://github.com/conda/conda/issues/7775#issuecomment-422426708
from @jamienoss
> Is there an issue/pr discussing the rationale for conda deciding not to publish gcc for OSX, that you could kindly point me to, please?
>
>So what about libgcc has that too been deprecated?
>
>For fp arithmetic, the order of operation can be important, in which case, consistency in compiler ?optimizations is also important. Being able to use gcc (compiler) on OSX helps mitigate these kinds of issues when the target arch/OS for the software is, for example, linux, i.e. scientific calibration pipeline production software running on a linux machine that the consumer (astronomer) may wish to re-use on a Mac.
>
>Of course, there is then the, perhaps more important, concern of using libgcc vs compiler-rt, but clang can be directed to use others, see LLVM doc here.

https://github.com/conda/conda/issues/7775#issuecomment-422434100
from @mingwandroid 
>Overlapping files in deprecated packages is fine by me, I've see no reason to attempt to fix deprecated things. Conda has no way of informing you of any news unfortunately. gcc was never really a user facing package from my perspective just something to get us out of a hole temporarily until we switched to clang.
>
>I don't know what you mean about macOS and Linux cross. Our gcc package never supported that being a native macOS compiler.
>
>If you want to create and use cross compilers you could look into how we construct gcc compilers these days. It's very cross happy being based upon crosstool-ng.
>
>We've blogged about the new compilers and conda-forge are well aware of them. Deprecation of gcc was only implicit in that blog post.

https://github.com/conda/conda/issues/7775#issuecomment-422435384
from @mingwandroid 
>The rationale for not providing gcc on macOS is simply that we provide only compilers that we use (for this is the primary purpose of them; to create the Anaconda Distribution) and macOS has moved to clang and for that reason so have we.
>
>The comment about math operation order needs elaboration, can you point to some bug demonstrating the problem?

https://github.com/conda/conda/issues/7775#issuecomment-422436037
from @mingwandroid 
>You should not expect exact results between OSes anyway due to libc differences.

**Comment by jamienoss** [comment]
> You should not expect exact results between OSes anyway due to libc differences.

You're right this is ultimately unavoidable due to architecture variance, however, the scope of this issue can be significantly reduced by using the same libraries across the various OSs (well at least OSX and Linux).

> The comment about math operation order needs elaboration, can you point to some bug demonstrating the problem?

Not quickly, at least not as an actual workable example. The bottom line is that language standards don't include optimizations and therefore compilers are not bound to be identical by compliance (to the language standard). That being said there are IEEE standards for FP math, i.e. IEEE 754, that include optimizations. E.g. the entire rationale behind the ``gcc`` option ``-funsafe-math-optimizations`` ([documented here](https://gcc.gnu.org/wiki/FloatingPointMath))

_"Due to roundoff errors the associative law of algebra do not necessarily hold for floating point numbers and thus expressions like (x + y) + z are not necessarily equal to x + (y + z)."_

It is possible for this to generate a difference in the accumulation of rounding errors due to evaluations being hoisted and then aggregated in different orders. Of course, using this option mitigates such a problem, however, only because GCC is IEEE 7 54 compliant. For the record, this option does exist for ``clang``.

I had a quick Google and couldn't find anything solidly stating that LLVM is, in fact, IEEE 754 compliant.

**Comment by jamienoss** [comment]
Breadcrumb: https://conda.io/docs/user-guide/tasks/build-packages/compiler-tools.html

**Comment by jamienoss** [comment]
@mingwandroid 
Are you able to provide a link to this blog, please?

The [above doc](https://conda.io/docs/user-guide/tasks/build-packages/compiler-tools.html) states that the switch defintaly happend with ``Anaconda 5.0``, I'm just wondering if the deprecationhappened earlier, when did it happen?

**Comment by jamienoss** [comment]
BTW: I doubt that LLVM (and thus clang) isn't IEEE 754 compliant, but perhaps this was discussed in the blog for deprecating GCC? It's definitely something that needs to be "ironed out".

**Comment by jamienoss** [comment]
Could you confirm whether ``libgcc`` has also been deprecated for OSX, please?

**Comment by jamienoss** [comment]
Found the following:
[5.0 release announcment](https://www.anaconda.com/blog/company-blog/announcing-the-release-of-anaconda-distribution-5-0/), but doesn't mention anything about deprecting anything.

However, these [5.0 release notes](http://docs.anaconda.com/anaconda/release-notes/#what-s-new-in-anaconda-5-0) do. They state that they (``libgcc`` and ``libgfortran `` ) have been "removed" for both 32bit and 64bit variations of linux, but nothing for OSX and nothing for the compiler ``gcc``.

**Comment by mingwandroid** [comment]
> Are you able to provide a link to this blog, please?

https://www.anaconda.com/blog/developer-blog/utilizing-the-new-compilers-in-anaconda-distribution-5/

>  It's definitely something that needs to be "ironed out".

You are free to use whatever compilers you wish, but the only one *we* support on `macOS` is `clang`. I don't *think* we ever recommended anyone to use the `gcc` package on `macOS`, in the past we used it only occasionally, and used the `macOS` 10.9 system compilers for most things. You are of course welcome to build your own `gcc` compiler stack and use that with `conda-build` if you wish, but if you use `gcc` then do not expect smooth interoperation with AD macOS packages and libraries! We need homogeneity here and not using `clang` on `macOS` is tilting at windmills IMHO.

> Could you confirm whether libgcc has also been deprecated for OSX, please?

It exists but we do not use it for anything in `AD` since 5.0. There is a `libgfortran-ng` (and `gfortran_osx-64-4.8.5`) package that *is* still used and it takes parts of the old `gcc` package.

You can see all of the 'main' `macOS` packages we've built since `AD` 5.0 here: https://repo.anaconda.com/pkgs/main/osx-64/
